src/compiler/checker.ts

```diff
@@ -5595,9 +5595,14 @@ namespace ts {
 
                 if (source.flags & TypeFlags.TypeParameter) {
                     let constraint = getConstraintOfTypeParameter(<TypeParameter>source);
+
                     if (!constraint || constraint.flags & TypeFlags.Any) {
                         constraint = emptyObjectType;
                     }
+
+                    // The constraint may need to be further instantiated with its 'this' type.
+                    constraint = getTypeWithThisArgument(constraint);

```

Comment:
You want to instantiate the constraint with `source` as the this-type parameter, so this line should be:

```Typescript
constraint = getTypeWithThisArgument(constraint, source);
```

As you have it now you're instantiating the constraint with its own this-type as the this-type parameter. I think the reason it works with the example in the bug is because the this-type of `Something` is assignable to the this-type of `Bar`, but it's not the right solution.